lib/BackbeatConsumer.js

```diff
@@ -361,19 +379,22 @@ class BackbeatConsumer extends EventEmitter {
 
     /**
      * Get how many messages we may attempt to consume at this time,
-     * considering the maximum concurrency, current processing queue
-     * length and message retrievals still pending.
+     * considering the maximum concurrency, queued messages, and message
+     * retrievals still pending.
      *
      * @return {integer} a strictly positive number representing the
      * number of new messages that the pipeline is ready to process,
      * or zero if the processing pipeline is 100% busy
      */
     _getAvailableSlotsInPipeline() {
-        const nSlots = this._concurrency
-              - this._processingQueue.running()
-              - this._processingQueue.length()
+        // do not exceed max queued tasks limit
+        const nQueuedSlots = this._maxQueued
+                - this._taskScheduler.length();
+        // do not exceed max concurrency limit
+        const nRunningSlots = this._concurrency
+              - this._taskScheduler.running()
               - this._nConsumePendingRequests;
-        return nSlots > 0 ? nSlots : 0;
+        return Math.max(0, Math.min(nQueuedSlots, nRunningSlots));
```

Comment:
would this not allow exceeding the concurrency, not sure that is what we want?

```suggestion
        // do not exceed max queued tasks limit
        const nQueuedSlots = this._maxQueued
                - this._taskScheduler.length();
        // do not exceed max concurrency limit
        const nRunningSlots = this._concurrency
              - this._taskScheduler.running()
              - this._taskScheduler.length()
              - this._nConsumePendingRequests;
        return Math.max(0, Math.min(nQueuedSlots, nRunningSlots));
```

or is that with `async.queues()`, once we 'started' a task it could be in the queue but not yet running : and so we had to compute both running() and length() ?

Reply:
Including `length()` in the calculation of concurrency slots won't really work here, as we may have configured a much larger "queuing factor".

Even if the tasks that need to be run are first "queued" for a brief moment this should still work as we always take the smallest value between concurrent slots and queued slots, so the tasks that we end up pushing fit in both the queue and concurrency slots.